parsers/US_ERCOT.py

```diff
@@ -0,0 +1,366 @@
+                decompressed = gzip.decompress(response.content)
+                zip_data = BytesIO(decompressed)
+            except gzip.BadGzipFile as err:
+                raise ValueError("File is neither a ZIP nor a gzipped file") from err
+
+        year_file = f"IntGenbyFuel{year}.xlsx"
+
+        with zipfile.ZipFile(zip_data) as zf:
+            if year_file not in zf.namelist():
+                raise NotImplementedError(
+                    f"Data for year {year} not found in historical data"
+                )
+            with zf.open(year_file) as excel_file:
+                return pd.read_excel(excel_file, engine="openpyxl", sheet_name=month)
+
+
+def fetch_historical_production(
+    zone_key: ZoneKey,
+    session: Session,
+    target_datetime: datetime,
+    logger: Logger = getLogger(__name__),
+) -> list:
+    if target_datetime.tzinfo is None:
+        target_datetime = target_datetime.replace(tzinfo=timezone.utc)
+
+    end = target_datetime + timedelta(hours=1)
+    start = end - timedelta(days=1)
+    year = target_datetime.year
+    month = target_datetime.strftime("%b")
+
+    all_data_points = []
+
+    df = get_sheet_from_date(year, month, session)
+
+    df["Date"] = pd.to_datetime(df["Date"])
+    time_columns = df.columns[4:]
+    datapoints_by_date = {}
+
+    for _, row in df.iterrows():
+        date = datetime.strptime(str(row["Date"]), "%Y-%m-%d %H:%M:%S").replace(
+            tzinfo=TX_TZ
+        )
+        if date < (start + timedelta(days=-1)) or date > (end + timedelta(days=1)):
+            continue
+
+        production_source = GENERATION_MAPPING[row["Fuel"]]
+
+        for hour in range(0, 24):
+            hour_dt = date + timedelta(hours=hour)
+
+            if hour_dt < start or hour_dt > end:
+                continue
+
+            # We get the data in 15 minute intervals, so we need to sum the 4 columns that represent an hour
+            start_idx = hour * 4
+            end_idx = start_idx + 4
+            hour_cols = time_columns[start_idx:end_idx]
+            hour_value = row[hour_cols].sum()
+
+            if hour_dt not in datapoints_by_date:
+                datapoints_by_date[hour_dt] = {"storage": {}, "production": {}}
+
+            target = "storage" if production_source == "battery" else "production"
+            datapoints_by_date[hour_dt][target].update({production_source: hour_value})
+
+    for hour_dt, production_and_storage in datapoints_by_date.items():
+        production = production_and_storage.get("production", {})
+        storage = production_and_storage.get("storage", {})
+        data_point = {
+            "zoneKey": zone_key,
+            "datetime": hour_dt,
+            "production": production,
+            "storage": storage,
+            "source": "ercot.com",
+        }
+        all_data_points.append(data_point)
+
+    return all_data_points
+
+
+def fetch_live_exchange(
+    zone_key1: str,
+    zone_key2: str,
+    session: Session = Session(),
```

Comment:
Generally we need to avoid setting mutable defaults, thous should default to None and then the session should be handled inside the function.